pkg/util/azureclient/applens/applens_client.go

```diff
@@ -0,0 +1,137 @@
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/runtime"
+)
+
+// AppLens client is used to interact with the Azure AppLens service.
+type Client struct {
+	endpoint string
+	pipeline runtime.Pipeline
+}
+
+// Endpoint used to create the client.
+func (c *Client) Endpoint() string {
+	return c.endpoint
+}
+
+// NewClient creates a new instance of AppLens client with Azure AD access token authentication. It uses the default pipeline configuration.
+// endpoint - The applens service endpoint to use.
+// cred - The credential used to authenticate with the applens service.
+// options - Optional AppLens client options.  Pass nil to accept default values.
+func NewClient(endpoint, scope string, cred azcore.TokenCredential, o *ClientOptions) (*Client, error) {
+	return &Client{endpoint: endpoint, pipeline: newPipeline([]policy.Policy{runtime.NewBearerTokenPolicy(cred, []string{scope}, nil), &appLensBearerTokenPolicy{}}, o)}, nil
+}
+
+func newPipeline(authPolicy []policy.Policy, options *ClientOptions) runtime.Pipeline {
+	if options == nil {
+		options = NewClientOptions()
+	}
+
+	return runtime.NewPipeline("applens", serviceLibVersion,
+		runtime.PipelineOptions{
+			PerCall:  []policy.Policy{},
+			PerRetry: authPolicy,
+		},
+		&options.ClientOptions)
+}
+
+//  ListDetectors obtains the list of detectors for a service from AppLens.
+// ctx - The context for the request.
+// o - Options for Read operation.
+func (c *Client) ListDetectors(
+	ctx context.Context,
+	o *ListDetectorsOptions) (*http.Response, error) {
+	if o == nil {
+		o = &ListDetectorsOptions{}
+	}
+
+	azResponse, err := c.sendPostRequest(
+		ctx,
+		o,
+		nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return azResponse, nil
+}
+
+// GetDetector obtains detector information from AppLens.
+// ctx - The context for the request.
+// o - Options for Read operation.
+func (c *Client) GetDetector(
+	ctx context.Context,
+	o *GetDetectorOptions) (*http.Response, error) {
+	if o == nil {
+		o = &GetDetectorOptions{}
+	}
+
+	azResponse, err := c.sendPostRequest(
+		ctx,
+		o,
+		nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return azResponse, nil
+}
+
+func (c *Client) sendPostRequest(
+	ctx context.Context,
+	requestOptions appLensRequestOptions,
+	requestEnricher func(*policy.Request)) (*http.Response, error) {
+	req, err := c.createRequest(ctx, http.MethodPost, requestOptions, requestEnricher)
+	if err != nil {
+		return nil, err
+	}
+
+	return c.executeAndEnsureSuccessResponse(req)
+}
+
+func (c *Client) createRequest(
+	ctx context.Context,
+	method string,
+	requestOptions appLensRequestOptions,
+	requestEnricher func(*policy.Request)) (*policy.Request, error) {
+	if requestOptions != nil {
+		header := requestOptions.toHeader()
+		ctx = policy.WithHTTPHeader(ctx, header)
```

Comment:
IMO, the headers that applens server required should be added here instead from `requestOptions`. In fact, if we do that, there is no need of `appLensRequestOptions` at all. Although I like your design to allow customization, I think it can be simplified for this client.

Reply:
nvm. I see that options is used to pass in the resource ID. fine :)